src/Components/Components/src/Reflection/MemberAssignment.cs

```diff
@@ -3,48 +3,136 @@
-            return dictionary.Values.SelectMany(p => p);
+            return new PropertyEnumerable(dictionary);
+        }
+
+        private static bool IsInheritedProperty(PropertyInfo property, object others)
+        {
+            if (others is PropertyInfo single)
+            {
+                return single.GetMethod?.GetBaseDefinition() == property.GetMethod?.GetBaseDefinition();
+            }
+
+            var many = (List<PropertyInfo>)others;
+            foreach (var other in CollectionsMarshal.AsSpan(many))
+            {
+                if (other.GetMethod?.GetBaseDefinition() == property.GetMethod?.GetBaseDefinition())
+                {
+                    return true;
+                }
+            }
+
+            return false;
+        }
+
+        public ref struct PropertyEnumerable
+        {
+            private readonly PropertyEnumerator _enumerator;
+
+            public PropertyEnumerable(Dictionary<string, object> dictionary)
+            {
+                _enumerator = new PropertyEnumerator(dictionary);
+            }
+
+            public PropertyEnumerator GetEnumerator() => _enumerator;
+        }
+
+        public ref struct PropertyEnumerator
+        {
+            // Do NOT make this readonly, or MoveNext will not work
+            private Dictionary<string, object>.Enumerator _dictionaryEnumerator;
+            private Span<PropertyInfo>.Enumerator _spanEnumerator;
+
+            public PropertyEnumerator(Dictionary<string, object> dictionary)
+            {
+                _dictionaryEnumerator = dictionary.GetEnumerator();
+                _spanEnumerator = Span<PropertyInfo>.Empty.GetEnumerator();
+            }
+
+            public PropertyInfo Current
+            {
+                get
+                {
+                    if (_dictionaryEnumerator.Current.Value is PropertyInfo property)
+                    {
+                        return property;
+                    }
+
+                    return _spanEnumerator.Current;
+                }
+            }
+
+            public bool MoveNext()
+            {
+                if (_spanEnumerator.MoveNext())
+                {
+                    return true;
+                }
+
+                if (!_dictionaryEnumerator.MoveNext())
+                {
+                    return false;
+                }
+
+                var oneOrMoreProperties = _dictionaryEnumerator.Current.Value;
+                if (oneOrMoreProperties is PropertyInfo)
+                {
+                    _spanEnumerator = Span<PropertyInfo>.Empty.GetEnumerator();
+                    return true;
+                }
+
+                var many = (List<PropertyInfo>)oneOrMoreProperties;
+                _spanEnumerator = CollectionsMarshal.AsSpan(many).GetEnumerator();
+                var moveNext = _spanEnumerator.MoveNext();
+                Debug.Assert(moveNext, "We expect this to at least have one item.");
+                return moveNext;
+            }
```

Comment:
This is a lot of extra machinery to reason about and maintain, especially with the caveats like "Do NOT make this readonly". People reading it may feel it could be done much more simply (e.g., the previous implementation) so it would be good to insert some kind of comment clarifying why we're doing all this.

How big is the payoff from doing all this? Bearing in mind that our plan is to move to a sourcegenerator-based way of setting component parameters, are we still going to be getting a meaningful benefit from doing this afterwards?